src/main/java/io/vertx/core/net/impl/ReloadingPemKeyManager.java

```diff
@@ -0,0 +1,244 @@
+/*
+* Copyright (c) 2011-2022 Contributors to the Eclipse Foundation
+*
+* This program and the accompanying materials are made available under the
+* terms of the Eclipse Public License 2.0 which is available at
+* http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+* which is available at https://www.apache.org/licenses/LICENSE-2.0.
+*
+* SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+*/
+package io.vertx.core.net.impl;
+
+import javax.net.ssl.ExtendedSSLSession;
+import javax.net.ssl.SNIServerName;
+import javax.net.ssl.SSLEngine;
+import javax.net.ssl.StandardConstants;
+import javax.net.ssl.X509ExtendedKeyManager;
+import javax.net.ssl.SNIHostName;
+
+import io.vertx.core.VertxException;
+import io.vertx.core.buffer.Buffer;
+import io.vertx.core.impl.VertxInternal;
+import io.vertx.core.impl.logging.Logger;
+import io.vertx.core.impl.logging.LoggerFactory;
+
+import java.net.Socket;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.nio.file.attribute.FileTime;
+import java.security.Principal;
+import java.security.PrivateKey;
+import java.security.cert.X509Certificate;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+public class ReloadingPemKeyManager extends X509ExtendedKeyManager {
+
+  private static final long FILE_POLLING_PERIOD_MSEC = 1000;
+  private static final Logger log = LoggerFactory.getLogger(ReloadingPemKeyManager.class);
+
+  private final VertxInternal vertx;
+  private CopyOnWriteArrayList<Credential> credentials = new CopyOnWriteArrayList<>();
+
+  public ReloadingPemKeyManager(VertxInternal vertx, List<String> certPaths, List<String> keyPaths,
+      List<Buffer> certValues, List<Buffer> keyValues) throws Exception {
+    this.vertx = vertx;
+
+    if ((keyPaths.size() < certPaths.size()) || (keyValues.size() < certValues.size())) {
+      throw new VertxException("Missing private key");
+    } else if ((keyPaths.size() > certPaths.size()) || (keyValues.size() > certValues.size())) {
+      throw new VertxException("Missing X.509 certificate");
+    } else if (keyPaths.isEmpty() && keyValues.isEmpty()) {
+      throw new VertxException("No credentials configured");
+    }
+
+    // Load credentials that were passed as file paths.
+    Iterator<String> cpi = certPaths.iterator();
+    Iterator<String> kpi = keyPaths.iterator();
+    while (cpi.hasNext() && kpi.hasNext()) {
+      credentials.add(new FileCredential(cpi.next(), kpi.next()));
+    }
+
+    // Create credentials that were passed by value.
+    Iterator<Buffer> cvi = certValues.iterator();
+    Iterator<Buffer> kvi = keyValues.iterator();
+    while (cvi.hasNext() && kvi.hasNext()) {
+      credentials.add(new Credential(cvi.next(), kvi.next()));
+    }
+
+    pollForFileChanges();
+  }
+
+  private void pollForFileChanges() {
+    vertx.setPeriodic(FILE_POLLING_PERIOD_MSEC, id -> {
```

Comment:
Why not use the JDK file watcher instead?

`FileSystems.getDefault().newWatchService()`

Reply:
I had another try before this PS with a `WatchService` and theads to run them https://github.com/eclipse-vertx/vert.x/compare/master...Nordix:certificate-rotation.

My problem there is that I cannot find a reliable way to delete the watch filehandles and threads. The `X509KeyManager` / `X509ExtendedKeyManager` interface have no close method and therefore they are not closed explicitly. I ran out of inotify filehandles when running test suite, since the key managers and watches were not garbage collected.  I'd appreciate any ideas here!

Now currently, I've been experimenting with another approach that avoids periodic polling:
When key or certificate is requested from key manager, return cached versions unless time-to-live period (of e.g. 1 second or something else) has passed. If cache TTL has passed, check file modify timestamp. If the files have been modified then reload.

Reply:
Maybe another (secondary) reason to not use `WatchService` is that they are sensitive on how the files are modified. In that previous link with my experiments, I added an example in comments how Kubernetes does the atomic swap of the files https://github.com/Nordix/vert.x/blob/db145f3ba4c6735a80fa005df1f30bdffd611b1a/src/main/java/io/vertx/core/net/impl/FileWatcher.java#L48-L85. One size does not fit all, so some configurability might need to be raised up all the way to all applications (through the frameworks that wrap Vert.x).